### ISSUE TITLE
It should be easier to fuzz nested data structures.

### ISSUE THREAD

**Issue by zenhack** [elm-explorations/test#100]
I'm in the midst of putting the finishing touches on a csv-encoding library before releasing it, and as part of that want a fuzzer for this type:

```elm
type alias Csv =
    { headers : List String
    , records : List (List String)
    }
```

The obvious way to do this would be something like:

```elm
csvFuzzer =
    Fuzz.map2 Csv
        (Fuzz.list Fuzz.string)
        (Fuzz.list (Fuzz.list Fuzz.string))
```

But this generates huge inputs, enough so that a simple test of "encode then decode gives the same result" takes >5 min on my laptop.

So I went about trying to write a custom fuzzer that generates more reasonably sized inputs. I've got something written that shrinks `headers` using `Shrink.list Shrink.string` and something that shrinks `records` using `Shrink.list (Shrink.list Shrink.string)`, and I have no idea how to combine them. If the LazyList module were exposed I could just concatenate the lists, but I have no idea how to do this with the exposed API.

There are a few possible takeaways from this experience:

* There are some gaps in the API for combining shrinkers
* You really shouldn't have to go to the trouble of writing custom generators and shrinkers just to get a smaller version of an existing fuzzer. Maybe something like #89 would help with this.
* It's awkward that a shrinker returns a type that isn't actually exposed anywhere, and thus can't be worked with.

What I would really have liked to write is something like this fuzzer:

```elm
csvFuzzer =
    Fuzz.map2 Csv
        (Fuzz.list Fuzz.string)
        (Fuzz.list (Fuzz.smaller (Fuzz.list Fuzz.string))
```

**Comment by mgold** [comment]
The issue with lazy lists is known and fixed on master.

Shrinking has no effect on the size of inputs tested. It only helps isolate failures. 

You will probably need a custom generator anyway so that you can generate random rows of random length. Use Shrink.noShrink as the shrinker.

**Comment by zenhack** [comment]
> Shrinking has no effect on the size of inputs tested. It only helps isolate failures.

Yes, I know -- the generation of smaller inputs in my experiments was done in the generator, but shrinking *is* important to the usability of the fuzzer, so `Shrink.noShrink` isn't a great solution.

I skimmed master, and it looks like LazyList has been hidden behind an opaque type, which makes things less confusing, but doesn't solve the problem that it's not obvious how to combine shrinkers (now called simplifiers).

I don't think I should need a custom generator in this case; while the one above will generate "ragged" csvs (with rows and the header being different lengths), it still works, catches the same issues, and is currently in use -- it's just really slow.

**Comment by zenhack** [comment]
Possibly related: https://github.com/elm-community/elm-test/issues/168

**Comment by Janiczek** [comment]
In 2.x.x there is a bunch of ways to control the list length (`listOfLengthBetween`, `listOfLength`, combinable with `andThen` as well).
